File: lambo/apps/websock.py
# -*- coding: utf-8 -*-  
import tornado.websocket 
import logging
import json 
def event(name):
    """Event handler decorator."""
    def handler(f):
        f._event_name = name
        return f

    return handler

class EventMeta(type):
    """Event handler metaclass"""
    def __init__(cls, name, bases, attrs): 
        events = {} 
        for a in attrs:
            attr = getattr(cls, a)
            name = getattr(attr, '_event_name', None)

            if name:
                events[name] = attr

        setattr(cls, '_events', events) 
        super(EventMeta, cls).__init__(name, bases, attrs)


class WebSocketEventHandler(tornado.websocket.WebSocketHandler):
    __metaclass__ = EventMeta
    
            
    def pack_event(self, name, data):
        msg = {'event': name,'data':data}
        return json.dumps(msg)
    
    def unpack_event(self, msg_json):
        return json.loads(msg_json)   
    
    def on_event(self,name,data):
        handler = self._events.get(name)
        if handler is not None:
            try:
                handler(self,data)
            except TypeError:
                logging.error('error try to call event handler %s with args %s'%(handler,repr(data)))
                raise
        else:
            logging.error('Invalid event name: %s' % name)
    
    def emit(self, name, data):
        msg = self.pack_event(name,data)
        self.write_message(msg) 
        
    def on_message(self, message): 
        msg = self.unpack_event(message)
        event = msg['event']
        data = msg['data']
        self.on_event(event,data)   
        
  

File: lambo/static/js/websock.js

/** websocket encapsulation **/

function pack_event(event,data){
	msg = { 'event': event, 'data': data } 
	return JSON.stringify(msg);
}
function unpack_event(message){
	return JSON.parse(message); 
}

function EventSocket(host){
	this.connect(host);
}

EventSocket.prototype.connect = function(host){
	var sock;
	if ("WebSocket" in window){
		sock = new WebSocket(host);
	} else if ("MozWebSocket" in window){
		sock = new MozWebSocket(host);
	}
	var $this = this;
	sock.onmessage = function(event){
		var msg = unpack_event(event.data); 
		var handler = $this.event_handler(msg.event); 
		if(handler){
			handler(msg.data);
		}
	};
	sock.onclose = function() {
		setTimeout(function() {
			$this.connect(host);
		}, 1000);
	};
	this.socket = sock;
	return this;
} 
EventSocket.prototype.emit = function(event,data){
	var msg = pack_event(event,data)
	this.socket.send(msg);
}

EventSocket.prototype.on = function(name,fn){
	if( !this.event_handlers ){
		this.event_handlers = {}
	}
	this.event_handlers[name] = fn;
}
EventSocket.prototype.event_handler = function(name){
	if(this.event_handlers){
		return this.event_handlers[name];
	}
	return null;
}
